components/brave_ads/browser/ads_p3a.cc

```diff
@@ -0,0 +1,142 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/browser/ads_p3a.h"
+
+#include <vector>
+
+#include "base/time/time.h"
+#include "brave/components/p3a/brave_p3a_service.h"
+#include "brave/components/p3a_utils/bucket.h"
+#include "brave/vendor/bat-native-ads/include/bat/ads/public/interfaces/ads.mojom-shared.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace brave_ads {
+namespace p3a {
+
+namespace {
+
+constexpr char kNewTabPageEventCountDictPref[] =
+    "brave_ads.p3a.ntp_event_count";
+constexpr char kViewCountKey[] = "views";
+constexpr char kClickCountKey[] = "clicks";
+constexpr char kExpireTimeKey[] = "expiry_time";
+
+constexpr int kCountBuckets[] = {0, 1, 2, 3, 8, 12, 16};
+
+constexpr size_t kHistogramExpiryTimeDays = 7;
+constexpr size_t kCountExpiryTimeDays = 30;
+
+std::string GetHistogramName(
+    const std::string& creative_instance_id,
+    const ads::mojom::NewTabPageAdEventType event_type) {
+  std::string result = "creativeInstanceId." + creative_instance_id;
+  switch (event_type) {
+    case ads::mojom::NewTabPageAdEventType::kServed:
+      NOTREACHED();
+      return std::string();
+    case ads::mojom::NewTabPageAdEventType::kViewed:
+      return result + ".views";
+    case ads::mojom::NewTabPageAdEventType::kClicked:
+      return result + ".clicks";
+  }
+}
+
+void UpdateMetricCount(PrefService* local_state,
+                       const std::string& pref_key,
+                       const std::string& histogram_name) {
+  DCHECK(local_state != nullptr);
+
+  DictionaryPrefUpdate update(local_state, kNewTabPageEventCountDictPref);
+  int count = 0;
+
+  base::Value* curr_value = update->FindPath({histogram_name, pref_key});
+  if (curr_value != nullptr && curr_value->is_int()) {
+    count = curr_value->GetInt();
+  }
+
+  count++;
+  update->SetPath({histogram_name, pref_key}, base::Value(count));
+  base::Time new_expiry_time =
+      base::Time::Now() + base::Days(kCountExpiryTimeDays);
+  update->SetPath({histogram_name, kExpireTimeKey},
```

Comment:
We should persist times using `TimeToValue` and `ValueToTime`, see `base/json/values_util`